src/rust/engine/fs/store/src/remote.rs

```diff
@@ -134,52 +135,56 @@ impl ByteStore {
+            req.set_resource_name(resource_name.clone());
+            req.set_write_offset(offset as i64);
+            let next_offset = min(offset + chunk_size_bytes, bytes.len());
+            req.set_finish_write(next_offset == bytes.len());
+            req.set_data(Bytes::from(&bytes[offset..next_offset]));
+            futures::future::ready(Some((
+              Ok((req, grpcio::WriteFlags::default())),
+              (next_offset, true),
+            )))
+          }
+        });
+
+        sender.send_all(&mut stream).await.or_else(move |e| {
+          match e {
+            // Some implementations of the remote execution API early-return if the blob has
+            // been concurrently uploaded by another client. In this case, they return a
+            // WriteResponse with a committed_size equal to the digest's entire size before
+            // closing the stream.
+            // Because the server then closes the stream, the client gets an RpcFinished
+            // error in this case. We ignore this, and will later on verify that the
+            // committed_size we received from the server is equal to the expected one. If
+            // these are not equal, the upload will be considered a failure at that point.
+            // Whether this type of response will become part of the official API is up for
+            // discussion: see
+            // https://groups.google.com/d/topic/remote-execution-apis/NXUe3ItCw68/discussion.
+            grpcio::Error::RpcFinished(None) => Ok(()),
+            e => Err(format!(
+              "Error attempting to upload digest {:?}: {:?}",
+              digest, e
+            )),
+          }
+        })?;
 
         sender
-          .send_all(stream)
-          .map(|_| ())
-          .or_else(move |e| {
-            match e {
-              // Some implementations of the remote execution API early-return if the blob has
-              // been concurrently uploaded by another client. In this case, they return a
-              // WriteResponse with a committed_size equal to the digest's entire size before
-              // closing the stream.
-              // Because the server then closes the stream, the client gets an RpcFinished
-              // error in this case. We ignore this, and will later on verify that the
-              // committed_size we received from the server is equal to the expected one. If
-              // these are not equal, the upload will be considered a failure at that point.
-              // Whether this type of response will become part of the official API is up for
-              // discussion: see
-              // https://groups.google.com/d/topic/remote-execution-apis/NXUe3ItCw68/discussion.
-              grpcio::Error::RpcFinished(None) => Ok(()),
-              e => Err(format!(
-                "Error attempting to upload digest {:?}: {:?}",
-                digest, e
-              )),
-            }
+          .close()
```

Comment:
A comment explaining the explicit close would be good.

Reply:
Done.